README.md
[![Travis CI status](https://api.travis-ci.org/obestwalter/py3status-pewpew.png)](https://travis-ci.org/obestwalter/py3status-pewpew)
[![PyPI version](https://badge.fury.io/py/py3status-pewpew.svg)](https://pypi.org/project/py3status-pewpew/)

# py3status-pewpew

A [py3status](https://github.com/ultrabug/py3status) module turning the [PewPew](https://pewpew.readthedocs.io) into a controller and external workspace display for the [i3wm](https://i3wm.org/) tiling window manager.

[![Watch a short demo video](http://img.youtube.com/vi/0Oy2CE2GZ7s/0.jpg)](http://www.youtube.com/watch?v=0Oy2CE2GZ7s "py3status-pewpew status controller")

Extend i3 with your PewPew. Use the buttons to send messages to i3 (or do whatever else you like) and always see which workspace you are on.

## install

```text
pip install py3status-pewpew
```

See install-ubuntu16.04.md for specific instructions on that distribution.

# testing

run the tests with tox:

```console
$ cd </path/to/this/repo>
$ tox
```

## development

install a development environment:

```console
$ cd </path/to/this/repo>
$ tox -e dev
```

... and activate it e.g via `source .tox/dev/bin/activate`. Then run tests with `pytest`.

Debugging hint - use screen to get direct access and run commands in the REPL:

    $ screen /dev/ttyACM0

This will open a screen session. Ctrl+C and Enter is useful to soft reboot pewpew in there.


CHANGELOG.md
# CHANGELOG

This project uses [smeantic versioning](https://semver.org/).

## [0.2] - 2019-08-09
## Added

* dependency to py3status >= 3.20 (we need the entry point discovery support) - thanks @objarni

## Fixed

* Adjust changelog format to use https://keepachangelog.com style

## [0.1.post1] - 2019-08-08
## Fixed 

* missing long description and homepage link for PyPI

## [0.1]  - 2019-08-08
## Added

* minimal implementation of a controller module for py3status using the PewPew

---

# About this Changelog

https://keepachangelog.com suggests:

Guiding Principles

* Changelogs are for humans, not machines.
* There should be an entry for every single version.
* The same types of changes should be grouped.
* Versions and sections should be linkable.
* The latest version comes first.
* The release date of each version is displayed.
* Mention whether you follow Semantic Versioning.

Types of changes

* **Added for new features.**
* **Changed for changes in existing functionality.**
* **Deprecated for soon-to-be removed features.**
* **Removed for now removed features.**
* **Fixed for any bug fixes.**
* **Security in case of vulnerabilities.**


tests/conftest.py
import pytest

from src.pew3wm.pew3wm import init_logging


@pytest.fixture(scope="session", autouse=True)
def setup_logging():
    init_logging()


tests/test_pew3wm.py
import logging

from pew3wm import pew3wm as mut


log = logging.getLogger(__name__)


def test_sanity():
    assert True


def test_path_finder():
    path = mut.find_pew_pew_path()
    print(f'Path is {path}')
    assert ['test'] == mut.find_pew_pew_path


tests/test_pewpew.py
import logging
from collections import namedtuple
from pkg_resources import parse_version

from py3status_pewpew.pewpew import (
    PewPewEvents,
    ACTION_EVENT,
    AXIS_EVENT,
    AXIS_HORIZONTAL,
    AXIS_VERTICAL,
    BUTTON_DOWN,
    BUTTON_K0,
    BUTTON_K1,
    MAJOR,
    MINOR,
)

event = namedtuple("event", ["type", "code", "value"])

log = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)


def test_pewpew_behaviour():
    class FakeParent:
        def __init__(self):
            self.py3 = {}

    said = []

    def say(something):
        said.append(something)

    parent = FakeParent()
    dut = PewPewEvents(parent, say)  # dut; Device Under Test
    dut.get_pewpew_device = lambda: "test"

    fake_events = [
        event(AXIS_EVENT, AXIS_HORIZONTAL, MINOR),
        event(AXIS_EVENT, AXIS_HORIZONTAL, MAJOR),
        event(AXIS_EVENT, AXIS_VERTICAL, MINOR),
        event(AXIS_EVENT, AXIS_VERTICAL, MINOR - 20),
        event(AXIS_EVENT, AXIS_VERTICAL, MAJOR),
        event(ACTION_EVENT, BUTTON_K0, BUTTON_DOWN),
        event(1, 2, 3),
        event(ACTION_EVENT, BUTTON_K1, BUTTON_DOWN),
    ]
    dut._get_events = lambda: fake_events
    dut.device = None
    dut._try()

    expected = [
        "Setting device to None",
        "Initialized",
        "Setting device to test",
        "Setting state to None",
        "Setting state to LEFT",
        "Setting state to RIGHT",
        "Setting state to UP",
        "Ignoring event (3, 1, -147)",
        "Setting state to DOWN",
        "Setting state to K0",
        "Ignoring event (1, 2, 3)",
        "Setting state to K1",
    ]
    print(said)
    assert said == expected


def test_py3status_version():
    import py3status.version
    assert parse_version(py3status.version.version) >= parse_version("3.20")


tools/code.py
import struct
import time

import pew
import supervisor
import usb_hid

pew.init()
screen = pew.Pix()
prev_state = 0

for i in range(1, 5):
    screen.pixel(0, 0, i)
    time.sleep(1)
    pew.show(screen)
screen.pixel(0, 0, 0)
pew.show(screen)

for gamepad in usb_hid.devices:
    if gamepad.usage_page == 0x01 and gamepad.usage == 0x05:
        break
else:
    raise RuntimeError("Gamepad HID device not found")
report = bytearray(6)

while True:
    if supervisor.runtime.serial_bytes_available:
        ux = input()
        if ux.isdigit():
            # pew.init()
            # screen = pew.Pix()
            pix = pew.Pix.from_text(ux, color=3)
            if int(ux) < 10:
                screen.blit(pix, 2, 2)
            else:
                screen.blit(pix, 1, 2)

    buttons = pew.keys()
    report_buttons = 0

    if buttons & pew.K_O:
        report_buttons |= 0x01

    if buttons & pew.K_X:
        report_buttons |= 0x02

    if buttons & pew.K_UP:
        y = -127
    elif buttons & pew.K_DOWN:
        y = 127
    else:
        y = 0

    if buttons & pew.K_LEFT:
        x = -127
    elif buttons & pew.K_RIGHT:
        x = 127
    else:
        x = 0

    # print(report)
    struct.pack_into("<Hbbbb", report, 0, report_buttons, x, y, 0, 0)
    gamepad.send_report(report)
    pew.show(screen)
    pew.tick(1 / 12)


tools/deploy.py
# coding: utf-8
import getpass
import grp
import logging
import os
import shutil

import distro
from pathlib import Path

log = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)
HERE = Path(__file__).parent
MODULE_NAME = "code.py"
PY3STATUS_MODULE_PATH = HERE.parent / "src/py3status_pewpew/pewpew.py"
PY3STATUS_CONFIG_PATH = ""


def main():
    warn_if_user_not_in_expected_groups()
    deploy_pew_pew_control_module()
    deploy_py3status_module()


def warn_if_user_not_in_expected_groups():
    expected_groups_by_distro = {"arch": ["uucp"], "ubuntu": ["input", "dialout"]}
    expected_groups = expected_groups_by_distro[distro.id()]
    user = getpass.getuser()
    groups = [g.gr_name for g in grp.getgrall() if user in g.gr_mem]
    for exp_group in expected_groups:
        if exp_group not in groups:
            log.info(f"warning: user '{user}' is not in expected group: '{exp_group}'")


def deploy_pew_pew_control_module():
    containerPath = find_container_path()
    if containerPath is None:
        raise Exception("Could not find pewpew; check cable!")
    assert containerPath.exists(), containerPath
    srcPath = HERE / MODULE_NAME
    dstPath = containerPath / MODULE_NAME
    print(f"deploying control module from {srcPath}...")
    if dstPath.exists():
        backupPath = dstPath.with_suffix(".bak")
        print(f"move existing {dstPath} to {backupPath}")
        dstPath.rename(backupPath)
    shutil.copy(srcPath, dstPath)
    assert srcPath.read_bytes() == dstPath.read_bytes(), "target != source; deploy fail?"
    print(f"...deployed control module to {dstPath}")


def deploy_py3status_module():
    configHomePath = Path(os.getenv("XDG_CONFIG_HOME", "~")).expanduser()
    candidates = [
        configHomePath / ".config/py3status/modules",
        configHomePath / ".config/i3status/py3status",
        configHomePath / ".config/i3/py3status",
        configHomePath / ".i3/py3status",
    ]
    path = first_existing(candidates)
    if path is not None:
        log.info(f"deploy py3status module {PY3STATUS_MODULE_PATH} to {path}")
        shutil.copy(PY3STATUS_MODULE_PATH, path)
    else:
        raise Exception(f"could not find a home for {MODULE_NAME}")


def first_existing(candidates):
    for candidate in candidates:
        print(f"checking {candidate}")
        if candidate.exists():
            return candidate
    return None


def find_container_path():
    user = getpass.getuser()
    candidates = [
        Path(f"/run/media/{user}/CIRCUITPY"),
        Path(f"/media/{user}/CIRCUITPY"),
    ]
    return first_existing(candidates)


if __name__ == "__main__":
    main()


src/pew3wm/listener.py
import evdev

ACTION_EVENT = 1
AXIS_EVENT = 3

BUTTON_K_0 = 304
BUTTON_K_1 = 305

AXIS_VERTICAL = 1
AXIS_HORIZONTAL = 0


def set_state(state):
    print(state)


devices = [evdev.InputDevice(fn) for fn in evdev.list_devices()]
for device in devices:
    if "PewPew" in device.name:
        break
    print(device.fn, device.name, device.phys)
else:
    raise Exception("No valid input")

device = evdev.InputDevice(device.path)


for event in device.read_loop():
    if event.type == AXIS_EVENT:
        if event.code == AXIS_HORIZONTAL:
            if event.value == -127:
                set_state("LEFT")
            elif event.value == 127:
                set_state("RIGHT")
        elif event.code == AXIS_VERTICAL:
            if event.value == -127:
                set_state("UP")
            elif event.value == 127:
                set_state("DOWN")
    elif event.type == ACTION_EVENT:
        if event.code == BUTTON_K_0 and event.value == 1:
            set_state("K0")
        elif event.code == BUTTON_K_1 and event.value == 1:
            set_state("K1")


src/pew3wm/pewpew.py
# -*- coding: utf-8 -*-
"""
Control i3wm with the PewPew (https://pewpew.readthedocs.io).

This py3status module displays the current i3 workspace
on a PewPew device, and make use of the buttons for various
i3 purposes. The default for buttons left and right
is to switch workspace, but all buttons are configurable.

Configuration parameters:
    button_down: List of commands to execute when DOWN button is pressed
        (default [])
    button_k0: List of commands to execute when K0 button is pressed
        (default [])
    button_k1: List of commands to execute when K1 button is pressed
        (default [])
    button_left: List of commands to execute when LEFT button is pressed
        (default ["i3-msg workspace previous_on_output"])
    button_right: List of commands to execute when RIGHT button is pressed
        (default ["i3-msg workspace next_on_output"])
    button_up: List of commands to execute when UP button is pressed
        (default [])
    cache_timeout: how often to (re)detect the current workspace (default 10)
    format: Format for module output (default "{state}")
    format_idle: output when PewPew is idle or disconnected (default "⎐")

Format placeholders:
    {state} name of the latest pressed button

@authors <the EuroPython py3status pewpew> <xxx>
@license BSD

Examples:
```
# display the pewpew as an icon
# show scratchpad on K0 press
# toggle the dpms py3status module by simulating a left click on it on K1 press
pewpew {
    format = "⎐"
    button_k0 = "i3-msg scratchpad show"
    button_k1 = "py3-cmd click 1 dpms"
}

# increase or decrease the volume on UP/DOWN press by simulating clicks on the
# volume_status module (if you have it loaded)
pewpew {
    button_up = "py3-cmd click 4 volume_status"
    button_down = "py3-cmd click 5 volume_status"
}
```

SAMPLE OUTPUT
{'color': '#00FF00', 'full_text': '⎐'}
"""
import logging

import json
import threading
import time

import evdev
import serial  # pyserial

ACTION_EVENT = 1
AXIS_EVENT = 3
AXIS_HORIZONTAL = 0
AXIS_VERTICAL = 1
BUTTON_DOWN = 1
BUTTON_K0 = 304
BUTTON_K1 = 305
MAJOR = 127
MINOR = -127

log = logging.getLogger(__name__)

SERIAL_DEVICE = "/dev/ttyACM0"

event_map = {
    (AXIS_EVENT, AXIS_HORIZONTAL, MINOR): "LEFT",
    (AXIS_EVENT, AXIS_HORIZONTAL, MAJOR): "RIGHT",
    (AXIS_EVENT, AXIS_VERTICAL, MINOR): "UP",
    (AXIS_EVENT, AXIS_VERTICAL, MAJOR): "DOWN",
    (ACTION_EVENT, BUTTON_K0, BUTTON_DOWN): "K0",
    (ACTION_EVENT, BUTTON_K1, BUTTON_DOWN): "K1",
}


class PewPewEvents(threading.Thread):

    state = None

    def __init__(self, parent, say=None):
        super(PewPewEvents, self).__init__()
        if say is not None:
            self._say = say

        self._set_device(None)
        self.parent = parent
        self._say("Initialized")

    # The 'say' function is here for automatic
    # testing purposes; see test_pewpew.py
    def _say(self, msg):
        pass

    def _set_device(self, device):
        self._say(f"Setting device to {device}")
        self.device = device

    def _set_state(self, state):
        self._say(f"Setting state to {state}")
        self.state = state
        self.parent.py3.update()

    def get_pewpew_device(self):
        devices = [evdev.InputDevice(fn) for fn in evdev.list_devices()]
        for device in devices:
            if "PewPew" in device.name and not 'event-mouse' in device.name:
                log.info("pewpew connected!")
                self._say("PewPew in device.name")
                break
        else:
            raise Exception("pewpew not found")
        return evdev.InputDevice(device.path)

    def _get_events(self):
        return list[self.device.read_loop()]

    def _try(self):
        try:
            if self.device is None:
                self._set_device(self.get_pewpew_device())
                self._set_state(None)

            for event in self._get_events():
                event_tuple = event.type, event.code, event.value
                new_state = event_map.get(event_tuple, None)
                if new_state is not None:
                    self._set_state(new_state)
                else:
                    self._say(f"Ignoring event {event_tuple}")

        except Exception:
            if self.state is not False:
                self._set_device(None)
                self._set_state(False)
                log.info("pewpew disconnected!")
            time.sleep(1)

    def run(self):
        while True:
            self._try()


class Py3status:

    # available configuration parameters
    button_down = []
    button_k0 = []
    button_k1 = []
    button_left = ["i3-msg workspace previous_on_output"]
    button_right = ["i3-msg workspace next_on_output"]
    button_up = []
    cache_timeout = 10
    format = "{state}"
    format_idle = "⎐"

    def post_config_hook(self):
        """
        This is the class constructor which will be executed once.
        """
        self._pewpew = PewPewEvents(self)
        self._pewpew.start()

    def _display_current_workspace(self):
        try:
            raw = self.py3.command_output("i3-msg -t get_workspaces")
            jsons = json.loads(raw)
            for j in jsons:
                if j["focused"]:
                    active = j["num"]
                    break
            else:
                active = 0
            # self.py3.log("active workspace is {}".format(active))
            with serial.Serial(SERIAL_DEVICE, 9600, timeout=1) as ser:
                ser.write(b"%d\r" % active)
        except Exception:
            pass

    def pewpew(self, *args, **kwargs):
        """
        """
        color = self.py3.COLOR_GOOD
        state = self._pewpew.state
        if state:
            button_exec = getattr(self, "button_{}".format(state.lower()), [])
            if isinstance(button_exec, str):
                button_exec = [button_exec]
            for cmd in button_exec:
                self.py3.command_run(cmd)
        elif state is False:
            state = self.format_idle
            color = self.py3.COLOR_BAD
        else:
            state = self.format_idle
        self._display_current_workspace()
        self._pewpew.state = None
        return {
            "cached_until": self.py3.time_in(self.cache_timeout),
            "color": color,
            "full_text": self.py3.safe_format(self.format, {"state": state}),
        }


if __name__ == "__main__":
    """
    Run module in test mode.
    """
    from py3status.module_test import module_test

    module_test(Py3status)
